candle-examples/examples/reinforcement-learning/ddpg.rs

```diff
@@ -0,0 +1,461 @@
+}
+impl Transition {
+    fn new(
+        state: &Tensor,
+        action: &Tensor,
+        reward: &Tensor,
+        next_state: &Tensor,
+        terminated: bool,
+        truncated: bool,
+    ) -> Self {
+        Self {
+            state: state.clone(),
+            action: action.clone(),
+            reward: reward.clone(),
+            next_state: next_state.clone(),
+            terminated,
+            truncated,
+        }
+    }
+}
+
+pub struct ReplayBuffer {
+    buffer: VecDeque<Transition>,
+    capacity: usize,
+    size: usize,
+}
+impl ReplayBuffer {
+    pub fn new(capacity: usize) -> Self {
+        Self {
+            buffer: VecDeque::with_capacity(capacity),
+            capacity,
+            size: 0,
+        }
+    }
+
+    pub fn push(
+        &mut self,
+        state: &Tensor,
+        action: &Tensor,
+        reward: &Tensor,
+        next_state: &Tensor,
+        terminated: bool,
+        truncated: bool,
+    ) {
+        if self.size == self.capacity {
+            self.buffer.pop_front();
+        } else {
+            self.size += 1;
+        }
+        self.buffer.push_back(Transition::new(
+            state, action, reward, next_state, terminated, truncated,
+        ));
+    }
+
+    pub fn random_batch(
+        &self,
+        batch_size: usize,
+    ) -> Option<(Tensor, Tensor, Tensor, Tensor, Vec<bool>, Vec<bool>)> {
+        if self.size < batch_size {
+            None
+        } else {
+            let transitions: Vec<Transition> = thread_rng()
+                .sample_iter(Uniform::from(0..self.size))
+                .take(batch_size)
+                .map(|i| self.buffer.get(i).unwrap().clone())
+                .collect();
+
+            let states: Vec<Tensor> = transitions
+                .iter()
+                .map(|t| t.state.clone().unsqueeze(0).unwrap())
+                .collect();
+            let actions: Vec<Tensor> = transitions
+                .iter()
+                .map(|t| t.action.clone().unsqueeze(0).unwrap())
+                .collect();
+            let rewards: Vec<Tensor> = transitions
+                .iter()
+                .map(|t| t.reward.clone().unsqueeze(0).unwrap())
+                .collect();
+            let next_states: Vec<Tensor> = transitions
+                .iter()
+                .map(|t| t.next_state.clone().unsqueeze(0).unwrap())
+                .collect();
+            let terminateds: Vec<bool> = transitions.iter().map(|t| t.terminated).collect();
+            let truncateds: Vec<bool> = transitions.iter().map(|t| t.truncated).collect();
+
+            Some((
+                Tensor::cat(&states, 0).unwrap(),
+                Tensor::cat(&actions, 0).unwrap(),
+                Tensor::cat(&rewards, 0).unwrap(),
+                Tensor::cat(&next_states, 0).unwrap(),
+                terminateds,
+                truncateds,
+            ))
+        }
+    }
+}
+
+fn track(
+    mut varmap: VarMap,
+    vb: VarBuilder,
```

Comment:
Passing this as a value rather than a ref makes it necessary to have a potentially large clone, could we just use references instead?